File: matiasmasso/Mat.NET/DT2/DTORanking.vb
Public Class DTORanking
    Property User As DTOUser
    Property Reps As List(Of DTORep)
    Property Rep As DTORep
    Property Proveidor As DTOContact
    Property FchTo As Date = DateTime.Today
    Property FchFrom As Date = New Date(DateTime.Today.Year, 1, 1)
    Property Days As Integer = 90
    Property Product As DTOProduct
    Property Area As DTOArea
    Property ConceptType As ConceptTypes
    Property Format As Formats
    Property Items As List(Of DTORankingItem)

    Property Proveidors As List(Of DTOContact)
    Property Catalog As DTOProductCatalog
    Property Channel As DTODistributionChannel

    Public Enum Formats
        Amounts
        Units
    End Enum

    Public Enum ConceptTypes
        Geo
        Product
    End Enum

    Shared Function Atlas(oRanking As DTORanking) As List(Of DTOCountry)
        Dim retval As New List(Of DTOCountry)
        Dim oCountries = oRanking.Items.GroupBy(Function(x) x.Location.Zona.Country.Guid).Select(Function(y) y.First).Select(Function(z) z.Location.Zona.Country).ToList
        Dim oZonas = oRanking.Items.GroupBy(Function(x) x.Location.Zona.Guid).Select(Function(y) y.First).Select(Function(z) z.Location.Zona).ToList
        Dim oLocations = oRanking.Items.GroupBy(Function(x) x.Location.Guid).Select(Function(y) y.First).Select(Function(z) z.Location).ToList

        For Each oCountry In oCountries
            retval.Add(oCountry)
            For Each oZona In oZonas.Where(Function(x) x.Country.Guid.Equals(oCountry.Guid))
                oCountry.Zonas.Add(oZona)
                For Each oLocation In oLocations.Where(Function(x) x.Zona.Guid.Equals(oZona.Guid))
                    oZona.Locations.Add(oLocation)
                Next
            Next
        Next
        Return retval
    End Function

    Shared Function CustomerRanking(oUser As DTOUser, Optional oProduct As DTOProduct = Nothing) As DTORanking
        Dim retval As New DTORanking
        With retval
            .User = oUser
            .ConceptType = DTORanking.ConceptTypes.Geo
            .Product = oProduct
        End With
        Return retval
    End Function

    Shared Function ExcelSheet(oRanking As DTORanking) As MatHelperStd.ExcelHelper.Sheet
        Dim retval As New MatHelperStd.ExcelHelper.Sheet()
        For Each item As DTORankingItem In oRanking.Items
            Dim oRow As MatHelperStd.ExcelHelper.Row = retval.AddRow()
            With item
                oRow.AddCell(item.Location.Zona.Country.ISO)
                oRow.AddCell(item.Location.Zona.nom)
                oRow.AddCell(item.Location.nom)
                oRow.AddCell(item.Customer.nom)
                Select Case oRanking.Format
                    Case DTORanking.Formats.Amounts
                        oRow.AddCell(item.Amt.Eur)
                    Case DTORanking.Formats.Units
                        oRow.AddCell(item.Units)
                End Select
            End With
        Next
        Return retval
    End Function

End Class

Public Class DTORankingItem
    Property Order As Integer
    Property Customer As DTOCustomer
    Property Location As DTOLocation
    Property Product As DTOProduct
    Property Units As Integer
    Property Amt As DTOAmt
End Class

File: matiasmasso/Mat.NET/DTO/DTORanking.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace DTO
{
    public class DTORanking
    {
        public DTOUser User { get; set; }
        public List<DTORep> Reps { get; set; }
        public DTORep Rep { get; set; }
        public DTOContact Proveidor { get; set; }
        //public DateTime FchTo { get; set; } = DTO.GlobalVariables.Today();
        //public DateTime FchFrom { get; set; } = new DateTime(DTO.GlobalVariables.Today().Year, 1, 1);
        public int Year { get; set; }
        public int MonthFrom { get; set; }
        public int MonthTo { get; set; }
        public int Days { get; set; } = 90;
        public DTOProduct Product { get; set; }
        public DTOArea Area { get; set; }
        public ConceptTypes ConceptType { get; set; }
        public Formats Format { get; set; }
        public List<DTORankingItem> Items { get; set; }

        public List<DTOContact> Proveidors { get; set; }
        public DTOProductCatalog Catalog { get; set; }
        public DTODistributionChannel Channel { get; set; }

        public bool GroupCcx { get; set; } // Group stores from same owner

        public enum Formats
        {
            Amounts,
            Units
        }

        public enum ConceptTypes
        {
            Geo,
            Product
        }

        

        public static List<DTOCountry> Atlas(DTORanking oRanking)
        {
            List<DTOCountry> retval = new List<DTOCountry>();
            var oCountries = oRanking.Items.GroupBy(x => x.Location.Zona.Country.Guid).Select(y => y.First()).Select(z => z.Location.Zona.Country).ToList();
            var oZonas = oRanking.Items.GroupBy(x => x.Location.Zona.Guid).Select(y => y.First()).Select(z => z.Location.Zona).ToList();
            var oLocations = oRanking.Items.GroupBy(x => x.Location.Guid).Select(y => y.First()).Select(z => z.Location).ToList();

            foreach (var oCountry in oCountries)
            {
                retval.Add(oCountry);
                foreach (var oZona in oZonas.Where(x => x.Country.Guid.Equals(oCountry.Guid)))
                {
                    oCountry.Zonas.Add(oZona);
                    foreach (var oLocation in oLocations.Where(x => x.Zona.Guid.Equals(oZona.Guid)))
                        oZona.Locations.Add(oLocation);
                }
            }
            return retval;
        }

        public static DTORanking CustomerRanking(DTOUser oUser, DTOProduct oProduct = null/* TODO Change to default(_) if this is not a reference type */)
        {
            DTORanking retval = new DTORanking();
            {
                var withBlock = retval;
                withBlock.User = oUser;
                withBlock.ConceptType = DTORanking.ConceptTypes.Geo;
                withBlock.Product = oProduct;
            }
            return retval;
        }

        public static MatHelper.Excel.Sheet ExcelSheet(DTORanking oRanking)
        {
            MatHelper.Excel.Sheet retval = new MatHelper.Excel.Sheet();
            foreach (DTORankingItem item in oRanking.Items)
            {
                MatHelper.Excel.Row oRow = retval.AddRow();
                {
                    var withBlock = item;
                    oRow.AddCell(item.Location.Zona.Country.ISO);
                    oRow.AddCell(item.Location.Zona.Nom);
                    oRow.AddCell(item.Location.Nom);
                    oRow.AddCell(item.Customer.Nom);
                    switch (oRanking.Format)
                    {
                        case DTORanking.Formats.Amounts:
                            {
                                oRow.AddCell(item.Amt.Eur);
                                break;
                            }

                        case DTORanking.Formats.Units:
                            {
                                oRow.AddCell(item.Units);
                                break;
                            }
                    }
                }
            }
            return retval;
        }
    }

    public class DTORankingItem
    {
        public int Order { get; set; }
        public DTOCustomer Customer { get; set; }
        public DTOLocation Location { get; set; }
        public DTOProduct Product { get; set; }
        public int Units { get; set; }
        public DTOAmt Amt { get; set; }
    }
}
